### PR TITLE
Cypress/E2E: Automate backlogs - Messaging > Custom Emoji (9 test cases)

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Custom Emoji  tests.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-28264

### CODE REVIEW

File: e2e/cypress/integration/emoji/custom_emoji_spec.js [hunk 1]
```diff
@@ -34,13 +80,520 @@ describe('Custom emojis', () => {
...
+        cy.getLastPostId().then((postId) => {
+            const postText = `#postMessageText_${postId}`;
+            cy.get(postText).should('have.text', messageText);
+
+            // # Press the up arrow to open the edit modal
+            cy.get('#post_textbox').type('{uparrow}');
+
+            // # Search for the pre-built emoji
+            // * Show emoji list
+            cy.get('#editPostEmoji').should('be.visible').click();
```
Comment (saturninoabril) [inline e2e/cypress/integration/emoji/custom_emoji_spec.js:446]: Step 4 is not via editing a post but reacting to a post -- "React again to that post by clicking emoji picker, typing the first three letters of the custom emoji name in search, and clicking the custom emoji". May open reaction to a post by,
```javascript
cy.clickPostReactionIcon(postId);
```
Screen Shot 2020-10-08 at 8 02 33 PM

File: e2e/cypress/integration/emoji/custom_emoji_spec.js [hunk 2]
```diff
@@ -34,13 +80,520 @@ describe('Custom emojis', () => {
...
+        // # Edit post and add a message
+        cy.getLastPostId().then((postId) => {
+            const postText = `#postMessageText_${postId}`;
+            cy.get(postText).should('have.text', messageText);
+
+            // # Edit the last post
+            cy.get('#post_textbox').type('{uparrow}');
+
+            // * Edit post modal should appear, and edit the post
+            cy.get('#editPostModal').should('be.visible');
```
Comment (saturninoabril) [inline e2e/cypress/integration/emoji/custom_emoji_spec.js:527]: Per test case [MM-T2187](https://automation-test-cases.vercel.app/test/MM-T2187), it should be done by post reaction and not by edit modal. Then, please verify if the last post has reaction.
Screen Shot 2020-10-08 at 8 16 52 PM

**Review by saturninoabril — CHANGES_REQUESTED** [review]
@catalintomai Thanks for these E2Es. Please see comments.

File: e2e/cypress/integration/emoji/custom_emoji_spec.js [hunk 3]
```diff
@@ -34,13 +80,520 @@ describe('Custom emojis', () => {
...
+        // # Check that the emoji image appears in the message - extract the image url and compafe with local image
+        cy.getLastPost().find('p').find('span > span').then((imageSpan) => {
+            cy.expect(imageSpan.attr('title')).to.equal(customEmojiWithColons);
+
+            // # Filter out the url from the css background property
+            // url("https://imageurl") => https://imageurl
+            const url = imageSpan.css('background-image').split('"')[1];
+
+            // # Verify that the emoji image i sthe correct one
+            cy.fixture(largeEmojiFile).then((overrideImage) => {
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply e2e/cypress/integration/emoji/custom_emoji_spec.js]: added the resized version - still puzzled how did the tests pass before on my box.

File: e2e/cypress/integration/emoji/custom_emoji_spec.js [hunk 4]
```diff
@@ -34,13 +80,520 @@ describe('Custom emojis', () => {
...
+        cy.get('.backstage-form__footer').within(($form) => {
+            // # Click on Save
+            cy.wrap($form).find('.btn-primary').click();
+        });
+
+        // # Navigate to a channel
+        cy.visit(`/${testTeam.name}/channels/town-square`);
+
+        // # Post a message
+        cy.postMessage(messageText);
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply e2e/cypress/integration/emoji/custom_emoji_spec.js]: @saturninoabril , yes, I find this test description a bit unclear: Step 3 mentions the reaction with an upper-case builtin emoji name (which does not work in the current codebase - should we open a bug for it?) While the expected result section only mentions "Emoji search is not case-sensitive" - while there is no step (other than 4 which deal with the custom emoji) that does emoji search. In here I've added code for the current behavior.

[the remark this answers is not in the document]
Comment (saturninoabril) [reply e2e/cypress/integration/emoji/custom_emoji_spec.js]: Ticket filed - https://mattermost.atlassian.net/browse/MM-29543. For step 3, I thought the expectation was that it should successfully react to the last post but let's how does the ticket be resolved and we'll update the test case as need.

**Review by saturninoabril — APPROVED** [review]
@catalintomai Thanks for updating! LGTM, tested and passed.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tests created are consistent with JM4T test cases
- All tests run Green

LGTM!

**Pull request merged**